File: typescipt/src/language/tables/typeDefinition.ts


export class TypeDefinition extends RootNode {
   public TypeName Name new(): =;
   public override string NodeName => Name.Value;

   public Array<VariableDefinition> Variables = list<VariableDefinition>(): new;

   private TypeDefinition(string name, SourceReference reference) {
     super(reference);
     Name.ParseName(name);
   }

   internal static parse(name: NodeName, reference: SourceReference): TypeDefinition {
     return new TypeDefinition(name.Name, reference);
   }

   public override parse(context: IParseLineContext): IParsableNode {
     let variableDefinition = VariableDefinition.parse(VariableSource.Parameters, context);
     if (variableDefinition != null) Variables.Add(variableDefinition);
     return this;
   }

   public override getChildren(): Array<INode> {
     return Variables;
   }

   protected override validate(context: IValidationContext): void {
   }

   public override validateTree(context: IValidationContext): void {
     using (context.CreateVariableScope()) {
       base.ValidateTree(context);
     }
   }
}

File: src/Lexy.Compiler/Language/Types/TypeDefinition.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Lexy.Compiler.Parser;

namespace Lexy.Compiler.Language.Types;

public class TypeDefinition : ComponentNode, ITypeDefinition, IHasNodeDependencies
{
    private readonly List<VariableDefinition> variables = new();

    public TypeName Name { get; } = new();
    public override string NodeName => Name.Value;

    public IReadOnlyList<VariableDefinition> Variables => variables;

    private TypeDefinition(string name, SourceReference reference) : base(reference)
    {
        Name.ParseName(name);
    }

    internal static TypeDefinition Parse(NodeName name, SourceReference reference)
    {
        return new TypeDefinition(name.Name, reference);
    }

    public override IParsableNode Parse(IParseLineContext context)
    {
        var variableDefinition = VariableDefinition.Parse(VariableSource.Parameters, context);
        if (variableDefinition != null) variables.Add(variableDefinition);
        return this;
    }

    public IEnumerable<IComponentNode> GetDependencies(IComponentNodeList componentNodes)
    {
        var dependencies = Variables.SelectMany(variable =>
            variable.Type is IHasNodeDependencies hasDependencies
            ? hasDependencies.GetDependencies(componentNodes)
            : Array.Empty<IComponentNode>());
        return dependencies;
    }

    public override IEnumerable<INode> GetChildren()
    {
        return Variables;
    }

    protected override void Validate(IValidationContext context)
    {
    }

    public override void ValidateTree(IValidationContext context)
    {
        using (context.CreateVariableScope())
        {
            base.ValidateTree(context);
        }
    }
}